driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java

```diff
@@ -320,83 +317,313 @@ private BsonDocument asGetMoreCommandDocument() {
         return document;
     }
 
-    private void initFromQueryResult(final QueryResult<T> queryResult) {
-        serverCursor = queryResult.getCursor();
+    @Nullable
+    private ServerCursor initFromQueryResult(final QueryResult<T> queryResult) {
         nextBatch = queryResult.getResults().isEmpty() ? null : queryResult.getResults();
         count += queryResult.getResults().size();
+        return queryResult.getCursor();
     }
 
-    private void initFromCommandResult(final BsonDocument getMoreCommandResultDocument) {
-        QueryResult<T> queryResult = getMoreCursorDocumentToQueryResult(getMoreCommandResultDocument.getDocument(CURSOR),
-                                                                        connectionSource.getServerDescription().getAddress());
+    @Nullable
+    private ServerCursor initFromCommandResult(final BsonDocument getMoreCommandResultDocument) {
+        QueryResult<T> queryResult = getMoreCursorDocumentToQueryResult(getMoreCommandResultDocument.getDocument(CURSOR), serverAddress);
         postBatchResumeToken = getPostBatchResumeTokenFromResponse(getMoreCommandResultDocument);
         operationTime = getMoreCommandResultDocument.getTimestamp(OPERATION_TIME, null);
-        initFromQueryResult(queryResult);
+        return initFromQueryResult(queryResult);
     }
 
     private boolean limitReached() {
         return Math.abs(limit) != 0 && count >= Math.abs(limit);
     }
 
-    private void killCursor() {
-        if (serverCursor != null && killCursorOnClose) {
+    private BsonDocument getPostBatchResumeTokenFromResponse(final BsonDocument result) {
+        BsonDocument cursor = result.getDocument(CURSOR, null);
+        if (cursor != null) {
+            return cursor.getDocument(POST_BATCH_RESUME_TOKEN, null);
+        }
+        return null;
+    }
+
+    /**
+     * This class maintains all resources that must be released in {@link QueryBatchCursor#close()}.
+     * It also implements a {@linkplain #doClose() deferred close action} such that it is totally ordered with other operations of
+     * {@link QueryBatchCursor} (methods {@link #tryStartOperation()}/{@link #endOperation()} must be used properly to enforce the order)
+     * despite the method {@link QueryBatchCursor#close()} being called concurrently with those operations.
+     * This total order induces the happens-before order.
+     * <p>
+     * The deferred close action does not violate externally observable idempotence of {@link QueryBatchCursor#close()},
+     * because {@link QueryBatchCursor#close()} is allowed to release resources "eventually".
+     * <p>
+     * Only methods explicitly documented as thread-safe are thread-safe,
+     * others are not and rely on the total order mentioned above.
+     */
+    @ThreadSafe
+    private static final class ResourceManager {
```

Comment:
Can we consider making this a non-static inner class and removing namespace property, and perhaps moving serverCursor back to the top-level?  I find it a bit confusing that a class that "maintains all resources that must be released" also contains these two properties, both of which were already available as properties of QueryBatchCursor.

Reply:
Update: I'm sure you already considered it, so perhaps start by explaining the rationale.  Perhaps it's because making it an inner class gives it unnecessary access to many other fields that it doesn't need access to?

Reply:
I think `serverCursor` should stay part of the `ResourceManager` specifically because it is effectively a representation of a releasable resource (the resource is server-side, so we cannot have the resource itself in the `ResourceManager`).

Making `ResourceManager` an inner class will indeed reduce the isolation, but I am not concerned with it much within a single source file, so will happily do the change, though this is indeed the reason I tend to prefer static nested classes to inner classes.

What is a problem I was afraid someone would point out, is the method `ResourceManager.serverApi` :), which seems completely out of place. This is something `QueryBatchCursor` may extract in the constructor store as its field, instead of calling `ResourceManager.serverApi`. However, when I tried doing this previously, I discovered there is a number of unit tests that start failing, because Spock fails to mock the `final ServerApi` class (which does not mean the class should not be final). Those tests require `ConnectionSource.getServerApi` to be mocked explicitly. I will look at this situation again.

Reply:
Just to be clear: I'm not necessarily advocating for this change.  I'm just asking you to consider it, and if you decide it's better to leave it as is I will be ok with that.

Reply:
Right. I think changing this is likely beneficial.